src/number_theory/number_field.lean

```diff
@@ -249,4 +250,124 @@ end
 
 end roots
 
+section bounded
+
+open finite_dimensional polynomial set
+
+variables (K : Type*) [field K] [number_field K]
+variables (A : Type*) [normed_field A] [is_alg_closed A] [normed_algebra ℚ A]
+
+lemma bdd_coeff_of_norm_le (B : ℝ) :
+  ∃ (C : ℤ), ∀ x : K, (∀ φ : K →+* A, ∥φ x∥ ≤ B) → ∀ i, |(minpoly ℚ x).coeff i| ≤ C :=
+begin
+  let S := finset.bUnion
+    (finset.product (finset.range (finrank ℚ K + 1)) (finset.range (finrank ℚ K + 1)))
+    (λ x, ( { B ^ (x.1 - x.2) * (x.1.choose x.2) } : finset ℝ)),
+  let C := nat.ceil (S.max' _),
+  use C,
+  intros x hφ,
+  have hx : is_integral ℚ x := is_separable.is_integral _ _,
+  have h_roots_bdd_minpoly : ∀ z ∈ (map (algebra_map ℚ A) (minpoly ℚ x)).roots, ∥z∥ ≤ B,
+  { intros z hz,
+    rsuffices ⟨φ, rfl⟩ : ∃ (φ : K →+* A), φ x = z, {exact hφ φ },
+    letI : char_zero A := char_zero_of_injective_algebra_map ((algebra_map ℚ _).injective),
+    rw [← set.mem_range, rat_range_eq_roots, mem_root_set_iff, aeval_def],
+    convert (mem_roots_map _).mp hz,
+    repeat { exact monic.ne_zero (minpoly.monic hx), }},
+  have h_bdd_degree : (minpoly ℚ x).nat_degree ≤ finrank ℚ K,
+  { refine le_of_eq_of_le (intermediate_field.adjoin.finrank hx).symm _,
+    exact ℚ⟮x⟯.to_subalgebra.to_submodule.finrank_le, },
+  intro i,
+  by_cases hi : i < finrank ℚ K + 1,
+  { suffices : B ^ ((minpoly ℚ x).nat_degree - i) * ((minpoly ℚ x).nat_degree.choose i) ≤ C,
+    { rw ← @rat.cast_le ℝ,
+      apply le_trans _ this,
+      convert coeff_le_of_roots_le i _ _ h_roots_bdd_minpoly using 1,
+      { rw [rat.cast_abs, coeff_map, norm_algebra_map', ← rat.norm_cast_real, real.norm_eq_abs], },
+      exacts [minpoly.monic hx, is_alg_closed.splits_codomain _], },
+    { apply le_trans _ (nat.le_ceil (S.max' _)),
+      refine finset.le_max' S _ _,
+      exact finset.mem_bUnion.mpr ⟨⟨(minpoly ℚ x).nat_degree, i⟩, finset.mem_product.mpr
+        ⟨finset.mem_range_succ_iff.mpr h_bdd_degree, finset.mem_range.mpr hi⟩,
+          finset.mem_singleton.mpr rfl⟩, }},
+  { rw [coeff_eq_zero_of_nat_degree_lt, _root_.abs_zero],
+    exact nat.cast_nonneg _, linarith, },
+  { exact finset.bUnion_nonempty.mpr
+    ⟨⟨0 , 0⟩, finset.mem_product.mpr ⟨finset.mem_range_succ_iff.mpr (zero_le _),
+      finset.mem_range_succ_iff.mpr (zero_le _)⟩, finset.singleton_nonempty _⟩, },
+end
```

Comment:
```suggestion
begin
  let S := finset.bUnion
    (finset.product (finset.range (finrank ℚ K + 1)) (finset.range (finrank ℚ K + 1)))
    (λ x, ( { B ^ (x.1 - x.2) * (x.1.choose x.2) } : finset ℝ)),
  have hS : S.nonempty := finset.bUnion_nonempty.2
    ⟨⟨0 , 0⟩, finset.mem_product.mpr ⟨finset.mem_range_succ_iff.mpr (zero_le _),
    finset.mem_range_succ_iff.mpr (zero_le _)⟩, finset.singleton_nonempty _⟩,
  let C := nat.ceil (S.max' hS),

  refine ⟨C, λ x hφ i, _⟩,
  have hx : is_integral ℚ x := is_separable.is_integral _ _,
  have h_roots_bdd_minpoly : ∀ z ∈ (map (algebra_map ℚ A) (minpoly ℚ x)).roots, ∥z∥ ≤ B,
  { intros z hz,
    rsuffices ⟨φ, rfl⟩ : ∃ (φ : K →+* A), φ x = z, { exact hφ φ },
    letI : char_zero A := char_zero_of_injective_algebra_map ((algebra_map ℚ _).injective),
    rw [← set.mem_range, rat_range_eq_roots, mem_root_set_iff, aeval_def],
    convert (mem_roots_map _).mp hz,
    repeat { exact monic.ne_zero (minpoly.monic hx) }},
  have h_bdd_degree : (minpoly ℚ x).nat_degree ≤ finrank ℚ K := le_of_eq_of_le
      (intermediate_field.adjoin.finrank hx).symm (ℚ⟮x⟯.to_subalgebra.to_submodule.finrank_le),
      
  by_cases hi : i < finrank ℚ K + 1,
  { suffices : B ^ ((minpoly ℚ x).nat_degree - i) * ((minpoly ℚ x).nat_degree.choose i) ≤ C,
    { rw ← @rat.cast_le ℝ,
      apply le_trans _ this,
      convert coeff_le_of_roots_le i _ _ h_roots_bdd_minpoly using 1,
      { rw [rat.cast_abs, coeff_map, norm_algebra_map', ← rat.norm_cast_real, real.norm_eq_abs] },
      exacts [minpoly.monic hx, is_alg_closed.splits_codomain _] },
    { apply le_trans _ (nat.le_ceil (S.max' _)),
      refine finset.le_max' S _ _,
      exact finset.mem_bUnion.mpr ⟨⟨(minpoly ℚ x).nat_degree, i⟩, finset.mem_product.mpr
        ⟨finset.mem_range_succ_iff.mpr h_bdd_degree, finset.mem_range.mpr hi⟩,
          finset.mem_singleton.mpr rfl⟩, }},
  { rw [coeff_eq_zero_of_nat_degree_lt, _root_.abs_zero],
    exact nat.cast_nonneg _, linarith }
end
```
I think it's more clear if you introduce `hS` at the beginning and then use it (I've also done some cosmetic changes). In any case the proof is mathematically non trivial, would you mind add some comments *in the proof* explaining what is going on?

Reply:
Well, in fact, I split the proofs into several lemmas. It does indeed make it easier to ingest. Two of the lemmas are purely about polynomials so I think I'll PR them first in another PR.

Reply:
Well, I am not quite happy with #16670 so far (but that's a discussion for the other PR), but I think the proofs in ```number_field.lean``` are now easier to follow. Still, I can add some comments if you think it is still needed.

Reply:
If this one depends on #16670 you should say it the first message, otherwise things are a little confusing.

Reply:
I have put the dependancy in the top description. Is there anything more that I should do?

Reply:
That's fine. I having a look at the other PR.